pkg/devfile/adapters/common/generic.go

```diff
@@ -0,0 +1,240 @@
+	"github.com/openshift/odo/pkg/log"
+	"github.com/openshift/odo/pkg/machineoutput"
+	"github.com/pkg/errors"
+	"io"
+	"k8s.io/klog"
+	"strings"
+)
+
+// ComponentInfoFactory defines a type for a function which creates a ComponentInfo based on the information provided by the specified DevfileCommand.
+// This is used by adapters to provide the proper ComponentInfo identifying which component (including supervisor) to target when executing the command.
+type ComponentInfoFactory func(command common.DevfileCommand) (ComponentInfo, error)
+
+// GenericAdapter provides common code that can be reused by adapters allowing them to focus on more specific behavior
+type GenericAdapter struct {
+	AdapterContext
+	client                   ExecClient
+	logger                   machineoutput.MachineEventLoggingClient
+	componentInfo            ComponentInfoFactory
+	supervisordComponentInfo ComponentInfoFactory
+}
+
+// NewGenericAdapter creates a new GenericAdapter instance based on the provided parameters
+func NewGenericAdapter(client ExecClient, logger machineoutput.MachineEventLoggingClient, context AdapterContext, ciFactory ComponentInfoFactory, supervisorFactory ComponentInfoFactory) GenericAdapter {
+	return GenericAdapter{
+		AdapterContext:           context,
+		client:                   client,
+		logger:                   logger,
+		componentInfo:            ciFactory,
+		supervisordComponentInfo: supervisorFactory,
+	}
+}
+
+// ExecuteCommand simply calls exec.ExecuteCommand using the GenericAdapter's client
+func (a GenericAdapter) ExecuteCommand(compInfo ComponentInfo, command []string, show bool, consoleOutputStdout *io.PipeWriter, consoleOutputStderr *io.PipeWriter) (err error) {
+	return ExecuteCommand(a.client, compInfo, command, show, consoleOutputStdout, consoleOutputStderr)
+}
+
+// ExecuteDevfileCommandSynchronously executes the devfile init, build and test command actions synchronously
+func (a GenericAdapter) ExecuteDevfileCommandSynchronously(command common.DevfileCommand, show bool) error {
+	exe := command.Exec
+	var setEnvVariable, cmdLine string
+	for _, envVar := range exe.Env {
+		setEnvVariable = setEnvVariable + fmt.Sprintf("%v=\"%v\" ", envVar.Name, envVar.Value)
+	}
+	if setEnvVariable == "" {
+		cmdLine = exe.CommandLine
+	} else {
+		cmdLine = setEnvVariable + "&& " + exe.CommandLine
+	}
+	// Change to the workdir and execute the command
+	var cmd executable
+	if exe.WorkingDir != "" {
+		// since we are using /bin/sh -c, the command needs to be within a single double quote instance, for example "cd /tmp && pwd"
+		cmd = executable{ShellExecutable, "-c", "cd " + exe.WorkingDir + " && " + cmdLine}
+	} else {
+		cmd = executable{ShellExecutable, "-c", cmdLine}
+	}
+	return a.Execute(command, show, []executable{cmd})
+}
+
+// DefaultCommands returns the devfile commands to execute based on the specified command options
+func DefaultCommands(debug, restart bool) []executable {
```

Comment:
Maybe DefaultRunCommands as I don't see any build actions

Reply:
This function doesn't exist anymore.